pkg/fuse/passfd.go

```diff
@@ -0,0 +1,329 @@
+/*
+ Copyright 2024 Juicedata Inc
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package fuse
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"syscall"
+
+	"k8s.io/klog"
+
+	"github.com/juicedata/juicefs-csi-driver/pkg/util"
+)
+
+type Fds struct {
+	globalMu sync.Mutex
+	basePath string
+	fds      map[string]*fd
+}
+
+var GlobalFds *Fds
+
+func InitGlobalFds(basePath string) {
+	GlobalFds = &Fds{
+		globalMu: sync.Mutex{},
+		basePath: basePath,
+		fds:      make(map[string]*fd),
+	}
+	if err := GlobalFds.ParseFuseFds(basePath); err != nil {
+		return
+	}
+}
+
+func InitTestFds() {
+	GlobalFds = &Fds{
+		globalMu: sync.Mutex{},
+		fds:      make(map[string]*fd),
+	}
+}
+
+func (fs *Fds) ParseFuseFds(basePath string) error {
+	klog.V(6).Infof("parse fuse fd in basePath %s", basePath)
+	entries, err := os.ReadDir(basePath)
+	if err != nil {
+		klog.Errorf("read dir %s: %s", basePath, err)
+		return err
+	}
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		subEntries, err := os.ReadDir(path.Join(basePath, entry.Name()))
+		if err != nil {
+			klog.Errorf("read dir %s: %s", basePath, err)
+			return err
+		}
+		for _, subEntry := range subEntries {
+			if strings.HasPrefix(subEntry.Name(), "fuse_fd_comm.") {
+				klog.V(6).Infof("parse fuse fd in %s", subEntry.Name())
+				fs.parseFuse(entry.Name(), path.Join(basePath, entry.Name(), subEntry.Name()))
+			}
+		}
+	}
+	return nil
+}
+
+func (fs *Fds) GetFdAddress(podHashVal string) string {
+	if f, ok := fs.fds[podHashVal]; ok {
+		return f.serverAddressInPod
+	}
+
+	address := path.Join("/tmp", podHashVal, "fuse_fd_csi_comm.sock")
+	addressInPod := path.Join("/tmp", "fuse_fd_csi_comm.sock")
+	// mkdir parent
+	_ = os.MkdirAll(path.Join("/tmp", podHashVal), 0777)
```

Comment:
this directory doesn't seem to be remove?

Reply:
fixed